### PR TITLE
feat(datatypes): implement VectorOp::take

### LINKED ISSUE #1054: Implement VectorOp::take
### What problem does the new feature solve?

`VectorOp` defines a set of basic compute operations.
https://github.com/GreptimeTeam/greptimedb/blob/a9c8584c987318a41edd7fe4e2ebc382538db7af/src/datatypes/src/vectors/operations.rs#L30-L31

This ticket requests to add the `take` method, like [this one in arrow](https://docs.rs/arrow/latest/arrow/compute/kernels/take/fn.take.html). It can construct a new vector out of an existing one, with given indices, which is useful in scenarios like sorting and repeating.

### What does the feature do?

Add a new method `take` to `VectorOp`

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by e1ijah1**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
Add a new method `take` to `VectorOp`

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
closes #1054

### CODE REVIEW

File: src/datatypes/src/vectors/constant.rs [hunk 1]
```diff
@@ -83,6 +83,28 @@ impl ConstantVector {
...
+            if let Value::UInt32(idx) = indices.get(i) {
+                if idx >= self.len() as u32 {
+                    return error::BadArrayAccessSnafu {
+                        index: idx as usize,
+                        size: indices.len(),
+                    }
+                    .fail();
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (e1ijah1) [reply src/datatypes/src/vectors/constant.rs]: fixed

**Comment by killme2008** [comment]
@v0y4g3r @evenyag @waynexia  PTAL

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/datatypes/src/vectors/constant.rs [hunk 2]
```diff
@@ -83,6 +84,34 @@ impl ConstantVector {
...
+                        idx < len,
+                        error::BadArrayAccessSnafu {
+                            index: idx as usize,
+                            size: indices.len()
+                        }
+                    );
+                }
+                Ok(())
+            })
+            .collect::<Result<Vec<_>>>()?;
```
Comment (v0y4g3r) [inline src/datatypes/src/vectors/constant.rs:107]: [nitpick]: maybe you can use [`arrow::compute::lt_scalar`](https://docs.rs/arrow/33.0.0/arrow/compute/kernels/comparison/fn.lt_scalar.html) here, something like:
```rs
        let indices_arr = indices.to_arrow_array();
        let arr = indices_arr.as_any().downcast_ref::<UInt32Array>().unwrap();
        ensure!(
            !arrow::compute::min_boolean(&arrow::compute::lt_scalar(arr, len).unwrap()).unwrap(),
            BadArrayAccessSnafu{...}
        );
```

File: src/datatypes/src/vectors/constant.rs [hunk 3]
```diff
@@ -83,6 +84,34 @@ impl ConstantVector {
...
+        }
+        let len = self.len() as u32;
+        indices
+            .iter_data()
+            .map(|idx| {
+                if let Some(idx) = idx {
+                    ensure!(
+                        idx < len,
+                        error::BadArrayAccessSnafu {
+                            index: idx as usize,
```
Comment (evenyag) [inline src/datatypes/src/vectors/constant.rs:100]: suggestion: We could cast idx to usize instead of casting len to u32
```suggestion
        let len = self.len();
        indices
            .iter_data()
            .map(|idx| {
                if let Some(idx) = idx {
                    ensure!(
                        idx as usize < len,
                        error::BadArrayAccessSnafu {
                            index: idx as usize,
```

File: src/datatypes/src/vectors/constant.rs [hunk 4]
```diff
@@ -83,6 +84,34 @@ impl ConstantVector {
...
+                            size: indices.len()
```
Comment (evenyag) [inline src/datatypes/src/vectors/constant.rs:101]: suggestion: Since we check `idx < len`
```suggestion
                            size: len
```

File: src/datatypes/src/vectors/constant.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/datatypes/src/vectors/constant.rs:107]: [suggestion]: to maintain behavior consistency, you can just panic here when accessing out-of-bound elements like [here](https://github.com/GreptimeTeam/greptimedb/pull/1115/files#diff-bfa87fb596051c52504631f7d95b437a2a4b368c4552759e64e3cb5bdda36a9aR23).

File: src/datatypes/src/vectors/constant.rs [hunk 5]
```diff
@@ -83,6 +84,34 @@ impl ConstantVector {
             self.length,
         )))
     }
+
+    pub(crate) fn take_vector(&self, indices: &UInt32Vector) -> Result<VectorRef> {
+        if indices.is_empty() {
+            return Ok(self.slice(0, 0));
+        }
```
Comment (evenyag) [inline src/datatypes/src/vectors/constant.rs:91]: correctness: If the `indices` slot is null, arrow's take returns a null value. So the length of the result is equal to `indices.len()`. However, we could return error for a constant vector when indices contain null since constant vector doesn't support null.
```
ensure!(indices.null_count() == 0, error);
```

File: src/datatypes/src/vectors/constant.rs (hunk 2 above)
Comment (e1ijah1) [reply src/datatypes/src/vectors/constant.rs]: Nice catch, thanks

File: src/datatypes/src/vectors/constant.rs (hunk 5 above)
Comment (e1ijah1) [reply src/datatypes/src/vectors/constant.rs:90]: Nice catch, thanks

File: src/datatypes/src/vectors/operations/take.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        let v = ConstantVector::new(Arc::new(Int32Vector::from_slice([111])), input_length);
+        let indices = UInt32Vector::from_slice(indices);
+        let out = v.take(&indices).unwrap();
+
+        assert!(out.is_const());
+        assert_eq!(expect_length, out.len());
+    }
+
+    #[test]
+    fn test_take_constant() {
```
Comment (evenyag) [inline src/datatypes/src/vectors/operations/take.rs:132]: Also test taking null index?

File: src/datatypes/src/vectors/constant.rs [hunk 7]
```diff
@@ -83,6 +83,37 @@ impl ConstantVector {
...
+        let indices_arr = arr.as_any().downcast_ref::<UInt32Array>().unwrap();
+        if !arrow::compute::min_boolean(
+            &arrow::compute::lt_scalar(indices_arr, len as u32).unwrap(),
+        )
+        .unwrap()
+        {
+            return Err(error::ArrowComputeSnafu.into_error(
+                arrow::error::ArrowError::ComputeError("Array index out of bounds".to_string()),
+            ));
+        }
```
Comment (evenyag) [inline src/datatypes/src/vectors/constant.rs:110]: Looks like this require allocating an additional boolean array and makes getting index out-of-bounded inconvenient.  I'd prefer to use the original version that using a loop. @v0y4g3r

Comment (v0y4g3r) [reply src/datatypes/src/vectors/constant.rs:110]: > Looks like this require allocating an additional boolean array and makes getting index out-of-bounded inconvenient. I'd prefer to use the original version that using a loop. @v0y4g3r

I think either approach is ok, arrow compute may have better performance with simd enabled but brings more memory footprint. 

BTW I suggest just panic on out-of-bound access instead return an error, like `take_indices` of `PrimitiveVector`s.

**Review by v0y4g3r — APPROVED** [review]
LGTM
